File: a-basic/src/main/java/org/example/domain/Customer.java
package org.example.domain;

import java.util.Date;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.OneToMany;
import javax.persistence.Table;

/**
 * Customer entity bean.
 */
@Entity
@Table(name="be_customer")
public class Customer extends BaseModel {

  /**
   * Convenience Finder for 'active record' style.
   */
  public static final Finder<Long,Customer> find = new Finder<>(Long.class, Customer.class);
  
  @Column(length=100)
  String name;

  Date registered;
  
  @Column(length=1000)
  String comments;
  
  @OneToMany(mappedBy="customer")
  List<Contact> contacts;

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public Date getRegistered() {
    return registered;
  }

  public void setRegistered(Date registered) {
    this.registered = registered;
  }

  public String getComments() {
    return comments;
  }

  public void setComments(String comments) {
    this.comments = comments;
  }

  public List<Contact> getContacts() {
    return contacts;
  }

  public void setContacts(List<Contact> contacts) {
    this.contacts = contacts;
  }
  
}

File: e-kotlin-maven/src/main/java/org/example/domain/Customer.kt
package org.example.domain;

import com.avaje.ebean.Model
import java.util.*
import javax.persistence.*
import javax.validation.constraints.Size

/**
 * Customer entity bean.
 */
@Entity
@Table(name = "be_customer")
public class Customer() : BaseModel() {

  companion object : Model.Find<Long, Customer>() {}

  public var inactive: Boolean = false;

  @Size(max = 100)
  public var name: String? = null;

  public var registered: Date? = null;

  @Size(max = 1000)
  public var comments: String? = null;

  @ManyToOne(cascade = arrayOf(CascadeType.ALL))
  public var billingAddress: Address? = null;

  @ManyToOne(cascade = arrayOf(CascadeType.ALL))
  public var shippingAddress: Address? = null;

  @OneToMany(mappedBy = "customer", cascade = arrayOf(CascadeType.PERSIST))
  public var contacts: MutableList<Contact> = ArrayList();

  constructor (name: String) : this() {
    this.name = name;
  }

  override public fun toString(): String {
    return "customer(id:$id name:$name)";
  }

  /**
   * Helper method to add a contact to the customer.
   */
  fun addContact(contact: Contact) {

    // setting the customer is automatically done when Ebean does
    // a cascade save from customer to contacts. 
    contact.customer = this;
    contacts.add(contact);
  }


}
